### DISCUSSION TITLE
DSP5020 Batterie laden

### DISCUSSION THREAD

**Discussion by creepystefan** [syssi/esphome-dps#19] [General]
Hey, sehr nettes Tool, 
Würde mir sowas ähnliches vorstellen nur etwas anders herum.
Solaranlage Einspeisung ins Netz.
Ohne Batterie.
Im Haus an der Steckdose soll ein Netzteil angesteckt werden, dahinter ein XY6020 (ähnlich DSP5020), allerdings soll der Strom erhöht werden wenn der Shelly 3EM einen negativen Wert erkennt. Und damit den dahinter hängenden Akku laden.
Ich habe mal ein paar Sachen bestellt und hoffe dass der XY6020 identisch ist mit dem DSP5020 von der seriellen Schnittstelle.
Hinter dem Akku hängt dann ein DCDC Wandler 24v (Batterie) auf 5v z.B. für raspberry und 12v für fritzbox.
Würde mich nochmals dann melden wenn alle verkabelt ist.

Hättest du Interesse mir zu helfen?

**Comment by syssi** [comment]
Na klar! Treib das Thema gerne an. Ich schaue dann, ob ich etwas beitragen kann.

**Comment by creepystefan** [comment]
hey super vielen dank.
leider habe ich von programmieren auch keine große Ahnung.
Habe erstmal herausgefunden dass es dafür extra eine App gibt
Sinilink mit dem passenden Modul Sinilink [XY-WFPOW](https://de.aliexpress.com/i/1005004365676396.html)
Leider finde ich nichts im zusammenhang mit dem Netzteil XY6020
Leider auch nicht das Register von dem Modbus RTU
Hast du dazu informationen?

**Comment by creepystefan** [comment]
Hey, also ich habe jetzt alles hier, dass mit dem XY-WFPOW hat sich glaub ich erstmal erledigt.
da er keine GPIOS zur freien verfügung stellt.
Ich habe mal das WT32-ETH01 geflashed und die GPIO´s 4 und 5 benutzt.
https://www.letscontrolit.com/forum/viewtopic.php?t=9134
leider bekomme ich keine Daten in mein Home Assistant.
Die Entitäten werden in Home Assistant angezeigt.
Wird vielleicht doch ein anderes Register benutzt?
Kann ich irgendwie das Register des XY6020 herausfinden? Logger?

esp32:
  board: wt32-eth01        angepasst auf meine situation
  framework:
    type: arduino

https://github.com/kbs3056/XY6015L_pyGUI

hier gibt es anscheinend jemanden der es über Python gemacht hat für das 6015, sollte nur die Stromstärke sein zum 6020

**Reply by syssi** [reply]
Verfolgst du das Thema noch? Mit etwas Glück finden wir alle relevanten Register in deiner verlinkten Implementierung.

**Reply by creepystefan** [reply]
Hey. Danke das du dich meldest.
Ja verfolge das Thema noch. Versuche heute über einen ESP32 Node. Der ESP32 kann unter EspHome kein Software emulierten Seriellen Port machen. Man muss die Hardware Anschlüsse (3x beim ESP32) verwenden. Ich versuche es nochmal mit dem anderen ESP32.
Gebe dir dann nochmal Bescheid.
Danke dir

**Reply by syssi** [reply]
Wieso bevorzugst du SoftwareSerial über Hardware-UART? Ein ESP32 hat eine GPIO-Matrix und routet die drei zur Verfügung stehenden Hardware-UARTs auf beliebige GPIOs.

**Reply by creepystefan** [reply]
Muss ich später mal schauen. Laut EspHome soll Software User nicht beim ESP32 funktionieren. Deswegen meine Idee den Hardware Uart zu benutzen.
Ich probiere mal später verschieden Kombinationen aus.
Software Uart ist doch das wenn man beliebige Gpios benutzt, oder liege ich da fällig falsch.
Habe auch einen 8266 rumliegen. Müsste aber irgendwie die 3,3V machen. Der ESP32 von AZ-Delivary kann ja auch 5V.
Würde ich beforzugen, da das XY6020 5v Versorgung am Anschluss ausgiebig.

**Reply by syssi** [reply]
Ja, du liegst bei folgendem Punkt falsch: Ein ESP32 hat eine GPIO-Matrix und routet die Hardware-UARTs auf jenen GPIO, auf welche du die `uart`-Komponente konfigurierst. Ein ESP8266 ist intern festverdrahtet. Hier musst du ausgewählte GPIOs nutzen, hinter welchen ein Hardware-UART verfügbar ist. Nutzt du andere GPIOs bemerkt ESPHome dies und lädt stattdessen einen Software-UART (welcher ressourcenintensiv ist und auch mal das ein oder andere Bit überhört).

**Reply by creepystefan** [reply]
ah ok super danke dir für die Erklärung, dann starte ich mal alles melde mich nachher

**Reply by creepystefan** [reply]
habe laut diesem Shema
https://drive.google.com/file/d/1do7tb6tHbTQNlPX8fVwQQnwxOe8gZbed/view
Uart2 probiert GPIO17 und GPIO16, leider nichts
GPIO 4 und 2 probiert leider nichts.
noch eine Idee?
Entsprechend habe ich auch 
esp32:
  board: nodemcu-32s
verwendet und tx_pin1 und rx_pin1 den GPIOS angepasst

**Comment by creepystefan** [comment]
AAAAHHHHHHHHHHHHH, Siehe da, beim Logging wird jetzt was angezeigt, zumindest das was auf dem Display eingestellt ist.
- id: uart_01
  baud_rate: 115200
  #baud_rate: 9600

**Comment by creepystefan** [comment]
[18:30:22][D][uart_debug:158]: >>> "\x01\x03\x00\x00\x00\r\x84\x0F"
[18:30:22][I][dps:063]: Status frame received
[18:30:22][D][text_sensor:064]: 'dps device model': Sending state 'DPS21'
[18:30:22][D][sensor:094]: 'dps voltage setting': Sending state 25.00000 V with 2 decimals of accuracy
[18:30:22][D][number:012]: 'dps voltage setting': Sending state 25.000000
[18:30:22][D][sensor:094]: 'dps current setting': Sending state 0.20900 A with 3 decimals of accuracy
[18:30:22][D][number:012]: 'dps current setting': Sending state 0.209000
[18:30:22][D][sensor:094]: 'dps output voltage': Sending state 25.00000 V with 2 decimals of accuracy
[18:30:22][D][sensor:094]: 'dps output current': Sending state 0.00100 A with 3 decimals of accuracy
[18:30:22][D][sensor:094]: 'dps output power': Sending state 0.02000 W with 2 decimals of accuracy
[18:30:22][D][sensor:094]: 'dps input voltage': Sending state 48.26000 V with 2 decimals of accuracy
[18:30:22][D][text_sensor:064]: 'dps protection status': Sending state 'Normal'
[18:30:22][D][sensor:094]: 'dps backlight brightness': Sending state 20.00000 % with 0 decimals of accuracy
[18:30:22][D][sensor:094]: 'dps firmware version': Sending state 1.50000  with 1 decimals of accuracy
[18:30:22][W][component:214]: Component modbus took a long time for an operation (0.14 s).
[18:30:22][W][component:215]: Components should block for at most 20-30ms.
[18:30:22][D][uart_debug:158]: <<< "\x01\x03\x1A\t\xC4\x00\xD1\t\xC4\x00\x01\x00\x02\x12\xDA\x00\f\x00\x00\x00\xFA\x00\x00\x00\x01\x00\x15\x00\x0F\x91}"

Output Schalter und der Key Lock funktionieren noch nicht.
Versionsanzeige läuft in 0,5 Schritten nach oben und bei 5 wieder auf 0, total crazy.

**Reply by syssi** [reply]
Das sieht doch schonmal super aus. Kannst du dein Setup bitte noch einmal beschrieben? Insbesondere Model/Typ des Gerätes, aus welchem gerade die Daten geliefert werden?

**Reply by creepystefan** [reply]
230V - 48V Netzteil
XY6020 Eingang Netzteil 
XY Ausgang soll LifePo4 24V kommen.
parallel dazu Mini Buck 0-36 auf Einstellbar 5V für Raspberry (bzw ich habe Rock Pi 4C+)
HomeAssist
parallel Mini Buck 0-36 Eingestellt auf 12V für z.B. FritzBox
Shelly Pro 3EM im Sicherungskasten.

Es soll dann wenn ein Minus Wert am Shelly erkannt wird die Spannung am XY auf Ladespannung angehoben werden.
und der Strom entsprechend der Watt Leistung im Verhältniss zur Ladespannung eingestellt werden.

Wenn der Shelly keinen Überschuss erkennt also einen Positiven wert anzeigt, dann auf 24.5V von mir aus.
und Leistung auf auf 1,5 A.
Wenn die Batterie sozusagen 26V (angenommen) dann wird über das Netzteil kein Strom gezogen.
Bereits ausprobiert, funktioniert super.
sobald die Batterie leer wird und auf 24.5V geht, dann kommen 1.5 A zur erhaltung für die 5V und 12V damit alles nicht ausgeht.

ESP32 Node von AZ-Delivery
https://drive.google.com/file/d/1do7tb6tHbTQNlPX8fVwQQnwxOe8gZbed/view
tx_pin1: GPIO4
rx_pin1: GPIO2
baud_rate: 115200

**Reply by creepystefan** [reply]
ESP32 UART2 GPIO 16 und 17 geht nicht. ausprobiert

**Reply by creepystefan** [reply]
MQTT vorhanden und der Shelly Pro 3EM ist die Version 2 mit dem rtp Protokoll dass über eine URL abgerufen werden kann
das wird meine nächste aufgabe dass ESP Home DSP über das rtp den Wert ausliest und zum steuern übernimmt.
dann ist HomeAssist und der MQTT nicht zwingend notwendig

**Reply by creepystefan** [reply]
Wenn dir langweilig ist könntest du die definierbaren M Speicher mit rein nehmen bei dir in der example.
Bei dem RD Tech DPS habe ich das bei dir in der Anleitung gefunden. Also die M0 - M9
M0 ist anscheinend das Einschaltregister des DPS.
1 und 2 weis ich gerade nicht
3 - 9 frei definierbar
da könnte ich dann die BatterieLadespannungen und Ladestrom und Ladeerhaltspannung und Leerlaufspannung einstellen.
und somit dann abrufen bei den verschiedenen zustände, dann könnte man auch schnell eine 12V batterie einstellen oder 36V
oder 48V (48V geht beim DPS5020 aber nicht da glaube Ladespannung bei 54V oder so liegt) deswegen der XY6020 60V

**Comment by creepystefan** [comment]
super danke für die info der Register, ich habe es mal auf mein Github reingesetzt, ich hoffe es macht dir nichts aus?
ich werde noch DPS in XY6020 ändern.
und mal schauen wie ich das mit den anderen Register hinbekomme.

**Reply by syssi** [reply]
Alles gut, solange du die Lizenz beibehälst. Bitte die `LICENSE` ins Wurzelverzeichnis kopieren oder den Projekt-Dialog von GitHub verwenden um die Lizenz zu setzen.

**Comment by creepystefan** [comment]
LICENSE erledigt

**Comment by creepystefan** [comment]
Hi
also habe den Namen komplett von DPS auf XY6020 angepasst.
das neue Log sieht jetzt auch besser aus mit den Werten die er ausgibt.
Den Schalter werde ich als nächstes in angriff nehmen.
ich hoffe der OnOff Schalter ist bei den Adressen dabei die du ausliehst.
melde mich dann nochmals

[23:59:59][D][uart_debug:114]: >>> 01:03:00:00:00:0D:84:0F
[23:59:59][I][xy6020:063]: Status frame received
[23:59:59][D][text_sensor:064]: 'xy6020 device model': Sending state 'xy60208'
[23:59:59][D][sensor:094]: 'xy6020 voltage setting': Sending state 24.00000 V with 2 decimals of accuracy
[23:59:59][D][number:012]: 'xy6020 voltage setting': Sending state 24.000000
[23:59:59][D][sensor:094]: 'xy6020 current setting': Sending state 1.99900 A with 3 decimals of accuracy
[23:59:59][D][number:012]: 'xy6020 current setting': Sending state 1.999000
[23:59:59][D][sensor:094]: 'xy6020 output voltage': Sending state 24.00000 V with 2 decimals of accuracy
[23:59:59][D][sensor:094]: 'xy6020 output current': Sending state 0.01200 A with 3 decimals of accuracy
[23:59:59][D][sensor:094]: 'xy6020 output power': Sending state 0.28000 W with 2 decimals of accuracy
[23:59:59][D][sensor:094]: 'xy6020 input voltage': Sending state 48.23000 V with 2 decimals of accuracy
[23:59:59][D][text_sensor:064]: 'xy6020 protection status': Sending state 'Normal'
[23:59:59][D][sensor:094]: 'xy6020 backlight brightness': Sending state 1620.00000 % with 0 decimals of accuracy
[23:59:59][D][sensor:094]: 'xy6020 firmware version': Sending state 4.00000  with 1 decimals of accuracy
[23:59:59][W][component:214]: Component modbus took a long time for an operation (0.13 s).
[23:59:59][W][component:215]: Components should block for at most 20-30ms.
[23:59:59][D][uart_debug:114]: <<< 01:03:1A:09:60:07:CF:09:60:00:0C:00:1C:12:D7:1E:DC:00:00:69:44:00:03:00:51:00:08:00:28:88:5B

**Comment by creepystefan** [comment]
so melde mich mal wieder
habe jetzt den ausleseblock erweitert, (bis ich das herausgefunden habe)
Firmware funktioniert jetzt
Auch einen internen Temperatur Sensor habe ich hinzugefügt.
werde mich dann mal an die M1 - M9 Funktionen machen
Modell geht leider noch nicht.
ON OFF is gerade schwierig da über den XY6020 mein Rock PI läuft wenn ich ausschalte is aus :-)

**Comment by creepystefan** [comment]
Hey
so habe jetzt die richtigen Blöcke für Key Lock und Power on/off herausgefunden.
der Status wird richtig gelesen und richtig ausgegeben.
leider funktioniert das mit dem schreiben in das register nicht.
könntest du mal bitte bei mir in XY6020.cpp schauen (bei Zeilen 100 etwas habe ich die Blöcke gekennzeichnet)
vielleicht findet sich ja ein weg.

**Reply by syssi** [reply]
> leider funktioniert das mit dem schreiben in das register nicht.

Was genau meinst du damit? "Power on/off" ist auch ein Schreibvorgang. ;-)

**Reply by creepystefan** [reply]
Komisch. Also wenn ich am steuerteil von dem ein und ausschalte erkennt er es in homeassist.
Wenn ich in HA den Knopf drücke, dann passiert nix. Und bei der nächsten UART Abfrage nimmt der Knopf in HA den Status wieder an den das steuerteil hat.
Ich kann heute Abend nochmal schauen.
Mir kommt es so vor als würde er das BIT im Register nicht setzen.
Und die Stromanzeige ist um eine Komma Stelle verschoben. Das bekomme ich aber hin.

**Reply by syssi** [reply]
Ein Log, welches per `uart.debug` den rohen Verkehr ins Log schreibt plus ein paar versuchte Schaltvorgänge wären hilfreich. Ich habe ein Idee woran es liegt und wuerde gerne einmal den Datenverkehr sehen.

**Reply by creepystefan** [reply]
Log ist in meinem Repo drin. wenn ich Keylock oder Output schalte kommt ein Fehler vom Modbus Error function 0x90
kommt das evtl weil ich den Auslesebereich erweitert habe?

**Comment by creepystefan** [comment]
Log (2) mehrmals hin und her geschalten
ich habe jetzt noch in Log (2) Run Time mit drin, funktioniert wunderbar
gibt es eine möglichkeit die Zeit hintereinander zu stellen?
also nicht Stunden und Minuten einzeln sonder nebeneinander?
ist der Block 20 und 22
wäre aber nur eine nebensache.

**Reply by syssi** [reply]
Wo finde ich besagtes Log?

**Comment by creepystefan** [comment]
hey, ich wollte mal nachfragen ob du dir dass schon angeschaut hast?
ich habe noch ein Dokument für die Register gefunden. vll hilft es ja
könntest du wenn möglich mir ein beispiel machen für register 0x50 dann könnte ich es auf die
anderen adressen anwenden?

**Reply by syssi** [reply]
Gibt es weiterhin konkrete Register mit denen du nicht arbeiten kannst?

**Comment by creepystefan** [comment]
Hallo, Schade dass du dich nicht mehr gemeldet hast.
habe jetzt eine Möglichkeit gefunden komplett über ESPHome zu gehen ohne CPP
Werde es mal über die Tage fertig stellen.
Wenn es dich interresiert, ich habe alles über Packages gemacht.
theoretisch wie du das mit dem testxxx.yaml machen wolltest.
dann braucht man nur noch die Yaml in github ändern und dann kann einfach den ESP neu flashen
vll hilft dir das.
grüße
bei mir wäre es dann die main.yaml

**Reply by syssi** [reply]
Hi! Das Jahresende ist immer ziemlich wild, da bleibt wenig Zeit für Seitenprojekte. Ich habe aber deinen Fortschritt verfolgt und hatte nur Erfolge und keine offenen Fragen mehr wahrgenommen.

**Comment by creepystefan** [comment]
Hey. Ich habe das mit dem Code von dir verworfen, habe es komplett in Mit ESPHome gebastelt. Mit dem Modbus Controller. Funktioniert super.
Das Gerät selber wäre die Main.yaml. Ist aber noch nicht alles drin.
Bei meinem Test.yaml versuche ich gerade über die Leistung vom Sicherungskasten den Strom am XY6020 zu regeln.
Das funktioniert nur irgendwie teilweise. Das blöde ist sobald die Leistung wieder ins Plus geht steigt meine Regelung aus. Und Strom geht auf 0 dann habe ich wieder zu Viel Leistung den ich rausschicke dann erkennt er es und regelt. Aber dann springt er dauernd extrem hin und her.

Wie hast du das bei dem Lazy-Limiter gerechnet. In Schritten oder immer neu berechnet?

**Reply by syssi** [reply]
Am besten arbeitet der Lazy-Limiter, wenn den Smartmeter auch negativen Bezug (also Einspeisung ins Netz) ausweisen kann. Hier ein Beispiel:

1. Dein Backofen möchte aufheizen und zieht 1000W.
2. Das Smartmeter sieht diesen Bezug und signalisiert ihn als "beziehe 1000W vom Netzanbieter".
3. Deine Steuerung sieht diesen Bezug und faehrt den Inverter hoch. Angenommen er kann 1000W leisten, dann laesst die Steuerung den Inverter nun 1000W produzieren.
4. Das Smartmeter signalisiert 0W Netzbezug. 
5. Die Steuerung muss sich nun merken / daran erinnern, dass die 0W nur zustande kommen, weil der Inverter gerade 1000W liefert. Es werden also weiterhin 1000W vom Inverter angefordert.
6. Der Backofen ist warm und will jetzt nur noch 200W.
7. Das Smartmeter signalisiert -800W (eine Einspeisung ans Netz von 800W aus dem Inverter).
8. Die Steuerung sieht die negativen Wert und passt die Anforderung an den Inverter an (1000W-800W=200W).

Diese Regelung funktioniert nur, wenn der Regler / die Steuerung moeglichst traege agiert. Waere die Regelung schon nach einer Sekunde frustriert, dass die Anforderung von 1000W an den Inverter den Bezug nicht auf 0 gesenkt hat und fordert deshalb 2000W vom Inverter (weil diese ja noch fehlen koennten), dann gerät dein System in Schwingung. Ich wuerde fuer den Anfang nur alle 30 Sekunden den aktuellen Netzbezug pruefen und daraufhin einen alternativen Bezug beim Inverter anfordern. Dann wirst du sehen, dass du zwar dem echten Bedarf hinterher laeufst aber Schwingungen vermeidest.

**Reply by creepystefan** [reply]
Jup danke der Aufklärung soweit habe ich es verstanden.
Habe mal Elektronik gelernt. Bin aber leider schon etwas raus.
Bei mir ist es genau umgekehrt.
Wenn negativ (also Zuviel eingespeist wird). Soll er am Xy6020 mit dem Strom entsprechen hoch gehen. Damit meine Batterie entsprechend dem Negativen wert geladen wird. Und der Überschuss an die Batterie gegeben wird.
Umrechnung von negativ auf positiv funktioniert super. Somit habe ich den negativen Wert als positiven zum hoch rechnen.
AUFBAU:
Balkonkraftwerk!
Shelly 3EM!
Saldierte Leistung wird am ESP erkannt.
An einer X-beliebigen Steckdose.
Ein Netzteil 48v auf den Xy6020
Dann zur Batterie.
Xy6020 24v Batterie nicht laden.
Xy6020 28v Batterie (24V lifepo4)
Wird geladen.
Über den Shelly wird negativer wert ermittelt und dem entsprechend bei 28v der Strom geregelt.
Da steig ich noch nicht ganz dahinter.

Ich probiere es mal langsamer laufen zu lassen. Bin da etwas zu schnell mit 0.1 Sekunde.

Gebe dir dann nochmal Bescheid

Danke dir

**Reply by creepystefan** [reply]
Meinen verwendeten Code habe ich gerade nicht hochgeladen. Habe den in meinem HomeAssistant nur drin.

Meine aktuelle Rechnung
Negativ zu positiv
Positiv zu negativ.

Den gewandelten positiven Wert.
(Überschuss)
Von Watt in Strom.

Den errechneten strom + den Aktuell eingestellten Strom am xy6020

Dann den neu errechneteten Strom an xy6020 senden

Da kommt er durcheinander.

Muss mal testen mit langsamer machen

**Comment by creepystefan** [comment]
hey ho.
also nach vielen versuchen läuft jetzt meine regelung.
hatte viele Fehler in der Rechnung und im Übergeben der Werte.
Code muss noch verfeinert werden.
aktuell funktionierender Code in bei mir unter Bat-Loader.yaml
danke dir

melde mich nochmals falls ich noch probleme habe.

**Comment by creepystefan** [comment]
so habe jetzt den Code etwas verfeinert.
Das mit den Packages funktioniert jetzt auch.
sollte jetzt damit klar kommen.
danke dir

**Comment by bandit7311** [comment]
@creepystefan  Hallo Stefan, tolles Projekt, tolle Hardware, wie klappts denn momentan? Welche Firmware Version haben deine XY6020L ?  Hast du irgendeinem Bug feststellen können, beim Umschalten von CC nach CV? Wenn ich den Main.yaml benutze, welche Funktionalitäten erhalte ich?  Solong.B

**Reply by creepystefan** [reply]
Hey Bandit,
Code läuft wieder.

Bitte erstelle doch eine Diskusion in meinem Profil unter XY6020

Momentan benutze ich den BAT Loader. Über einen Shelly 3EM pro wird die Leistung im Sicherungskasten gemessen. Und bei einer PV Überschuss wird der Strom hochgeregelt dass ich wieder auf 0 Watt bin. Überschuss in die Batterie Egal an welcher Steckdose.
Somit kann ich noch nichts genaues sagen da die Sonne noch etwas fehlt.
Die Main Yaml hat fast alle standart Funktionen des XY6020.
Du kannst dir mal das Dokument anschauen. Habe versucht alle Funktionen aus der Liste zu implementieren.
Würdest du es gerne versuchen wollen?
Ich könnte dir dabei helfen, falls Interesse besteht.
Grüße 

Die Main.yaml ist standart Funktion ohne irgendwelche Regelung. Damit kannst du nur rein den XY6020 steuern. Also Spannung Strom. Und soweiter wie auf dem Standart Bediendisplay das beim Kauf dabei ist.

Die Bat-Loader.yaml ist da anders diese ist mit Regelung. Und müsste dann auch speziell angepasst werden.

**Comment by SberkChristian** [comment]
hi,
@creepystefan 
bin per Zufall auf die Diskussion gestoßen und das Thema hat großes Interesse bei mir geweckt. Leider verstehe ich von alldem viel zu wenig.

dürfte ich dir direkt ein paar Fragen stellen? Allerdings weiß ich nicht, wie ich mit dir in Kontakt treten könnte.

Viele Grüße

**Reply by creepystefan** [reply]
Hey.
Klar gerne kann ich weiter helfen.
Hast du bereits Erfahrung in EspHome? Wäre vorteilhaft.
Gehe bitte auf meine GitHub Seite und dann in den Repos dann unter esphome-xy6020 schreibe dort eine Diskussion oder eine issue.
Grüße
